Add furniController.lua

drucifer-sc opened the pull request:
Cheap fix to stop cascading storage updates from eating up all the working time available after failing exports, this needs to be fleshed out better, if the chest is full, scan the slots for any that match and are not full, stop blindly exporting.

Added furniController to miloApps, a multi-furnace-array controller, read the header notes for information

drucifer-sc committed:
Update exportTask.lua

Cheap fix to stop cascading storage updates from eating up all the working time available after failing exports, this needs to be more fleshed out better, if the chest is full, scan the slots for any that match and are not full, stop blindly exporting

drucifer-sc committed:
Merge pull request #1 from cynagen/export-patch-1

Update exportTask.lua

drucifer-sc committed:
Create furniController.lua

(Super)MultiFurnace Controller (did I mention, it's SUPER!?)
This app is designed to wrap up to 4 of kepler155c's MultiFurnace (multiFurni) arrays and divide up workload, it functions as a straight drop-in for furni.lua and no other adjustments are required to make it work. Additionally, you get the ability to target any one multiFurni array for smelting when creating the machine recipe in Milo. Further uses for this direct targeting can be limited-scope auto-smelting, using anywhere from 1/2 to 1/4 of your total capacity to auto-smelt, leaving the rest of the cluster available for you when you want it, great for multi-user bases!

Setup: Plug this turtle into a modem on your Milo network, and another modem leading to 2-4 turtles running furni.lua (or even furniController.lua for chained setups), configure useSlot4 on line 20, map in Milo and go!

drucifer-sc committed:
Merge pull request #2 from cynagen/furniController

Create furniController.lua

kepler155c commented:
This looks good to me - Anavrins - let me know if you have any comments.

I don't have much time to test out Milo changes. But if you get a chance, would you be able to test this change in storage.lua ? I saw that while reviewing your perf change.

Line 518 of storage.lua (basically - why try pushing other slots if the first one failed ?)
```
if amount > 0 then
	amount = transfer(key, amount, slot)
	if amount > 0 then
		source.lastUpdate = os.clock()
		target.lastUpdate = os.clock()
	else
		break
	end
end
`

drucifer-sc committed:
Update furniController.lua

Cleaned up bad reference naming to avoid confusion. Additionally moved relevant variables to local scope, cleaned up a few other minor things.

drucifer-sc commented:
Actually trying to avoid the cache update on line 528 altogether (which if I'm reading the code correctly, precedes the fix you're asking me to try), figured it'd be easier accomplished in the exportTask plugin in the first place, saving erroneous calls to export.

kepler155c commented:
I pushed a fix to address the export performance - the logic in storage.lua was pushing way too many times when the target was full. Therefore, we should not need the proposed change to exportTask.lua. Please review the commit I pushed for this fix.

drucifer-sc commented:
I applied your update of the storage API, and removed my fix from the affected Milo. The export task average time went straight from ~50ms with my fix only to ~1500ms, but is no longer growing exponentially. I still firmly believe that exportTask needs to be reworked in more detail to prevent unnecessary calls to Storage:export in the first place avoiding these perf hits altogether. I recognize my fix is in the wrong place and is very vague, I will see about reworking it myself.

drucifer-sc committed:
Update exportTask.lua

Fleshed out the exportItems function which was blindly firing items into full chests. Made it a little more sensitive, slight increase in average processing time, but no more erroneous hits to export causing perf issues.

cynagen commented on milo/plugins/exportTask.lua:
```diff
@@ -69,10 +69,25 @@ function ExportTask:cycle(context)
-							if node.adapter.size() ~= Util.size(node.adapter.list()) and context.storage:export(node, nil, item.count, item) == 0 then
-								-- TODO: really shouldn't break here as there may be room in other slots (probably not)
-								-- leaving for now for performance reasons
-								break
```
Bye bye shitty fix

kepler155c commented:
With the change i put in, the total execution time should not be hitting 1500. :(

The execution time should be calculated as:
50ms * number of unique items you are exporting to that single chest

If we are exceeding that amount, then possibly we are activating Squid's cost system multiplier (I have no idea how the cost system works).

Going through the storage/caching code is not an issue - it's extremely fast - the biggest factor is waiting for the plethora reply for every request (50ms per call).

I'll try out the change you provided - and I'll do some testing myself.

drucifer-sc commented:
Bear in mind for testing, I'm currently exporting to all slots [\*] 21-22 unique items on a standard-sized chest (27-slot). This is likely what's causing the problem in the first place as it's looping through trying to export some of each of those 22 items in it's filter list. There's also another 4 chests with single-items being exported to again, all slots [\*], so 25-26 individual items being exported, the majority against a single chest however. (I don't care what goes in when, just that it goes in eventually.)

drucifer-sc committed:
Update exportTask.lua

Clean up cached .list() after done, accidentally left it in, bloating storage config sizes

drucifer-sc committed:
Update exportTask.lua

Fixing some logic, make less plethora calls

drucifer-sc committed:
Update exportTask.lua

drucifer-sc committed:
Update exportTask.lua

Last fix, was making a call to Util when it wasn't defined, and left the export function completely dead. Defined correctly, now exporting as expected, whoops.

drucifer-sc commented:
Made my last update to exportTask, should button up the exports. With my fix in place, export times settled in around 200ms idle for 5 chests, which I guess is to be expected? The only qualm I have right now is my use of `node.cacheList` and I'm contemplating moving it to `local cache` instead to avoid contaminating the storage config at all.


Ran quick perf test comparison:
Before your `storage.lua` fix: Exponentially scaling proc time
Clean install of Milo (rm /sys, pastebin run uzghlbnc), with your fix: ~1400-1500ms proc time
My last version of `exportTask.lua` plus your storage api fix: 200ms or less idle; Lookin' good now!

Actually very broken, would not export smaller portions to fill, only large chunks of whatever was available at the time until all slots were populated, then fail due to bad usage of `filterItem`. Fixed the code in another revision, and times shot up. I suspect it's all the plethora calls I'm making.

drucifer-sc committed:
Update exportTask.lua

So uhh, yeah, the looped getItemMeta calls cost me a LOT of time. Figured out most of the relevant data I wanted was already being provided elsewhere, no extra calls to make, it's as fast as I can make it now. Now it's dependent on the storage optimizations as they stand, this does not speed anything up, just makes it more accurate.

kepler155c commented:
I pushed another change (exportTask.lua) - basically using your concept - please pull this file and give it a shot. Thanks

drucifer-sc commented:
Much better, exporter task is now sub-100ms with your fix. Might I suggest to increase the accuracy with regards to nbt, that when exporting an item with nbt, on a slot that matches the name and damage, *then* execute an .getItemMeta() on the slot to check for a matching nbt? Might be worth the little bit of a spike in activity for those unique items, since it'd only be those items. Now how do I retract my edit from the PR?

kepler155c commented:
Thanks for looking into that perf issue - I knew it was bad - but did uncover a lot of extra calls. I think what we have will work for +90% of the cases. The biggest issue is if they are exporting something with unique nbts and the chest is pretty full - this could generate getItemMetas on 20+ slots which will eat up a lot of time - best to just try and push and let that 1 call fail.

To undo the exportTask.lua - do a git checkout [FILE] - that will replace what you have with the current version. Then, just commit again.

drucifer-sc committed:
Sync with kepler155c's branch

drucifer-sc commented:
Maybe an option in the future to confirm NBT at cost of performance? Maybe an auto-disable recommendation if the process gets really heavy, say proc time of >2s or >97% of total time?

xAnavrins commented:
0619eee41c31accea72f3c6ee661c1feb4d873f5 broke autocrafting, items would get placed in the inventory and almost imediately get sucked back into the storage.
Removing the break on line 534 fixes it, but that kind of defeats the purpose of removing unessesary peripheral calls though.

kepler155c commented:
I reverted the break -- not sure why that broke anything though ... but need to keep autocrafting working

kepler155c commented:
Another update to exportTask.lua - I seem to have forgotten that list() does provide NBT info (I must have been thinking about 1.7 or turtles). Hopefully this is it.

drucifer-sc committed:
Another resync

drucifer-sc commented:
This seems to be it. All fixes in place, the system is ~50ms idle on exporter now, spiking up only when there's actual activity.
